File: ios/LLMHub/Sources/LLMHub/MimoBot/WhisperSTT.swift
import Foundation

/// whisper.cpp-backed `SpeechToText`.
///
/// NOT IMPLEMENTED YET — iOS v0 uses `SFSpeechRecognizerSTT`. When this lands,
/// it'll consume PCM from a `MicSource` (or the BLE transport) and run
/// whisper.cpp inference via a SwiftPM dependency.
///
/// TODO(whisper-ios):
///   1. `https://github.com/ggerganov/whisper.cpp` — the `whisper.spm` branch
///      builds cleanly on iOS 17+.
///   2. Transcribe once the mic source reports end-of-utterance (VAD or PTT).
final class WhisperSTT: SpeechToText {
    let modelURL: URL
    init(modelURL: URL) { self.modelURL = modelURL }

    func recognizeTurn(languageHint: String) async -> String {
        fatalError("TODO(whisper-ios): whisper.cpp init + transcribe")
    }

    func cancel() { /* TODO */ }
    func close() { /* TODO */ }
}

File: android/app/src/main/java/com/llmhub/llmhub/mimobot/speech/WhisperStt.kt
package com.llmhub.llmhub.mimobot.speech

/**
 * whisper.cpp-backed [SpeechToText].
 *
 * NOT IMPLEMENTED YET — the Android ships against a system recognizer for v0
 * (see [AndroidSpeechRecognizerStt]). When wired in, this implementation will:
 *   1. Own its own [com.llmhub.llmhub.mimobot.audio.MicSource] OR consume
 *      frames from the BLE transport.
 *   2. Buffer PCM until the VAD tail fires.
 *   3. Feed the buffer into whisper.cpp via JNI.
 *
 * TODO(whisper):
 *   1. Bundle libwhisper.so (build from https://github.com/ggerganov/whisper.cpp
 *      examples/whisper.android/lib, arm64-v8a + armeabi-v7a).
 *   2. Ship whisper-tiny.en-q5_1.bin (~75 MB) via the existing ModelDownloader.
 *   3. On [recognizeTurn], start mic + VAD; on end-of-utterance, call
 *      `WhisperContext.fullTranscribe(pcm)` and return the concatenated segments.
 */
class WhisperStt(
    @Suppress("unused") private val modelPath: String,
) : SpeechToText {

    override suspend fun recognizeTurn(languageHint: String): String {
        TODO("whisper.cpp JNI — load model, mic+VAD, whisper_full, join segments")
    }

    override fun cancel() { /* TODO */ }
    override fun close() { /* TODO */ }
}
